packages/create-docusaurus/src/index.ts

```diff
@@ -17,13 +16,20 @@ import supportsColor from 'supports-color';
 const RecommendedTemplate = 'classic';
 const TypeScriptTemplateSuffix = '-typescript';
 
-function hasYarn() {
-  try {
-    execSync('yarnpkg --version', {stdio: 'ignore'});
-    return true;
-  } catch (e) {
-    return false;
+const SupportedPackageManagers = ['npm', 'yarn', 'pnpm'];
+
+function getPackageManagerForUse(
+  forceUseNpm?: boolean,
+): typeof SupportedPackageManagers[number] {
+  if (forceUseNpm) {
+    return 'npm';
   }
+
+  return (
+    SupportedPackageManagers.find((packageManager) =>
+      process.env.npm_config_user_agent?.startsWith(packageManager),
+    ) || 'npm'
```

Comment:
At least, I don't want to use NPM as fallback but rather use the heuristic we had before. For example, I do have `create-docusaurus` installed globally, and if I run `create-docusaurus` without a package manager, it defaults to NPM with no way for me to select Yarn.

I think we can have three layers of heuristics:

1. Detect lockfiles. For example, the existence of a `pnpm-lock.yaml` or `package-lock.json` in a parent/grandparent folder is indicative enough that we are installing in an existing project.
2. If this is a standalone project, use the `npm_config_user_agent` except when the command is initiated by `npx`.
3. If the user has `create-docusaurus` installed globally, or if the command is using `npx`, we look for `yarnpkg` in the PATH.

Reply:
I agree that lock files do need to be considered, but I'm still not sure about using yarn as the default option. If Docusaurus should play equally well with any popular package manager, then using npm as default seems like the most reasonable or predictable as default option. This is well illustrated by the CRA, on whose behavior Docusaurus is based right now. However, since latest major release of CRA, they started to use yarn if end user initialized the new project via it, in all other cases npm is used (including npx). That's why I suppose we also need to adjust the current process of initializing in similar way -- it's more reasonable and expectable.

Reply:
Oh, I wasn't following the latest CRA, but interesting to know that they default to NPM now... In that case, I think maybe we should also add an optional `--pkg-manager` option and replace `--use-npm` with that, otherwise the user can only halt the installation and manually install again in case the pkg manager detection is not good? I'm trying to push back on NPM because I do have some personal opinions about it😅

Reply:
An option like the `--pkg-manager` might be a good idea, but in my opinion it would require more effort to implement, so within this PR I would only improve "detection mechanism" for package manager. `--use-npm` is more like a temporary workaround that may no longer be needed after this PR.

> I do have some personal opinions about it

Haha, I understand it, I have sort of a similar feeling, but we need to consider the interests of users first.